### PR TITLE
cli: fix `cockroach quit`

### LINKED ISSUE #25870: cli: `cockroach quit` reports transient RPC error during server shutdown, confuses acceptance tests
During a merge to master build for #25863 (which doesn't seem to have related changes), `TestDockerCLI/test_audit_log.tcl` flaked: https://teamcity.cockroachdb.com/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=675599&_focus=1873

There seems to be an error encountered during shutdown of the test.  The end of the output is:

```
.180523 18:29:23.153390098 EXPECT TEST: INTERRUPT TO FOREGROUND PROCESS

Error: interrupted
Failed running "sql"

.180523 18:29:23.561740353 EXPECT TEST: BEGIN STOP SERVER
graceful shutdown failed: rpc error: code = Internal desc = grpc: compressed flag set with identity or empty encoding; proceeding with hard shutdown
Error: unable to connect or connection lost.

Please check the address and credentials such as certificates (if attempting to
communicate with a secure cluster).

hard shutdown failed: Failed to connect to the node: error sending drain request: rpc error: code = Canceled desc = grpc: the client connection is closing
Failed running "quit"
child process exited abnormally
    while executing
"system "$argv quit""
    (procedure "stop_server" line 4)
    invoked from within
"stop_server $argv"
    (file "/go/src/github.com/cockroachdb/cockroach/cli/interactive_tests/test_audit_log.tcl" line 87)
non-zero exit code: 1
    	cli_test.go:92: non-zero exit code: 1
```

### LINKED ISSUE #26144: roachtest: decommission/nodes=4/duration=1h0m0s failed on master
SHA: https://github.com/cockroachdb/cockroach/commits/83e98d996b3ba3537afd7bc67e439eb5d3ea1659

Parameters:

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=683712&tab=buildLog

```
	decommission.go:218,decommission.go:235: /home/agent/work/.go/bin/roachprod run teamcity-683712-decommission-nodes-4-duration-1h0m0s:1 -- ./cockroach quit --insecure --port {pgport:1}: exit status 1

```

### PR DESCRIPTION
**Pull request by knz**
Informs/fixes #25870.
Fixes  #26144.

This patch fixes the following:

- the logic in `doShutdown()` aims to ignore errors caused by attempts
  connect to a server which is closing its gRPC channels, but was
  missing one case of such errors: during the initial check whether
  the node was running. This patch causes gRPC "closed connection"
  errors to become also ignored in that case.

- previously if there was a transient gRPC error during a hard
  shutdown whereby the shutdown could still succeed, then `cockroach
  quit` would fail no matter what. This patch makes `cockroach quit`
  retry a hard shutdown.

- the warning messages are now emitted on stderr (via `log.Warningf`)
  instead of stdout.

Release note (bug fix): fix a bug where `cockroach quit` would
erroneously fail even though the node already successfully shut down.

Release note (cli change): `cockroach quit` now emits warning message
on its standard error stream, not standard output.

### CODE REVIEW

**Comment by tschottdorf** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[pkg/cli/start.go, line 1075 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26158#-LDfcD_k3H_qC5OZ1duQ:-LDfcD_lCaxUYReHXFO5:b-75963u) ([raw file](https://github.com/cockroachdb/cockroach/blob/23238cb66bb43197560bbc9b8213ad5362a345ce/pkg/cli/start.go#L1075)):*
> ```Go
> 				return nil
> 			}
> 			if strings.Contains(err.Error(), "proceeding with hard shutdown") {
> ```

Could you extract this string constant so that it's shared with the other use? As is, it will rot too easily.

---

*[pkg/cli/start.go, line 1078 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26158#-LDfcKLG6v1_0EtSaQdR:-LDfcKLH8DGlWEq-dl33:b-928nvn) ([raw file](https://github.com/cockroachdb/cockroach/blob/23238cb66bb43197560bbc9b8213ad5362a345ce/pkg/cli/start.go#L1078)):*
> ```Go
> 				// Something happened server-side that made the server decide on
> 				// a hard shutdown of its own. This is not an error, but still
> 				// inform the user.
> ```

Perhaps also mention that we don't fold this into `IsClosedConnection` because we don't know whether this error can only occur in that circumstance (don't think so).

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26158#-:-LDfcVD-9tlxMCuBqY_T:bnfp4nl)*

**Comment by knz** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/cli/start.go, line 1075 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26158#-LDfcD_k3H_qC5OZ1duQ:-LDfeo8L6a8tJins625W:b-38fq9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/23238cb66bb43197560bbc9b8213ad5362a345ce/pkg/cli/start.go#L1075)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Could you extract this string constant so that it's shared with the other use? As is, it will rot too easily.
</blockquote></details>

I made a mistake -- this string is not generated by the server. Solved the problem in a different way (retry hard shutdown).

---

*[pkg/cli/start.go, line 1078 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26158#-LDfcKLG6v1_0EtSaQdR:-LDfeurRATevikk_4cFE:bb7k4oj) ([raw file](https://github.com/cockroachdb/cockroach/blob/23238cb66bb43197560bbc9b8213ad5362a345ce/pkg/cli/start.go#L1078)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Perhaps also mention that we don't fold this into `IsClosedConnection` because we don't know whether this error can only occur in that circumstance (don't think so).
</blockquote></details>

n/a.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26158)*

**Comment by tschottdorf** [comment]
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26158)*

**Comment by knz** [comment]
bors failed on #26098, retrying

bors r+

**Comment by knz** [comment]
(bors was stuck)

bors r+

File: pkg/cli/start.go [hunk 1]
```diff
@@ -1045,6 +1045,9 @@ func doShutdown(ctx context.Context, c serverpb.AdminClient, onModes []int32) er
 	// out, or perhaps drops the connection while waiting). To that end, we first
 	// run a noop DrainRequest. If that fails, we give up.
 	if err := checkNodeRunning(ctx, c); err != nil {
+		if grpcutil.IsClosedConnection(err) {
```
Comment (a-robinson) [inline pkg/cli/start.go]: As @bdarnell asked on https://github.com/cockroachdb/cockroach/pull/26163, does this swallow the error if you run `quit` against a server that isn't running?

Comment (knz) [reply pkg/cli/start.go]: answered to ben there. if the server does not exist the thing still fails.

**Comment by couchand** [comment]
internal bors weirdness, retrying

bors r=knz

**Pull request merged**